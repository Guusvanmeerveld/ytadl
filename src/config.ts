import { z } from 'zod';

const CRON_REGEX = /(@(annually|yearly|monthly|weekly|daily|hourly|reboot))|(@every (\d+(ns|us|Âµs|ms|s|m|h))+)|((((\d+,)+\d+|(\d+(\/|-)\d+)|\d+|\*) ?){5,7})/;

export const streamOptions = z.object({
	audio: z
		.object({
			included: z.boolean().optional(),
		})
		.optional(),
	video: z
		.object({
			included: z.boolean().optional(),
		})
		.optional(),
});

export const configItem = z.object({
	name: z.string().optional(),
	id: z.string().optional(),
	cron: z.string().regex(CRON_REGEX).optional(),
	platform: z.string().default('youtube'),
});

export default z.array(configItem);
