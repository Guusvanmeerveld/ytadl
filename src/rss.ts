import fetch from 'node-fetch';

import { YoutubeFeed } from './interfaces/feed';

export const fetchFeed = async (feed: string): Promise<YoutubeFeed> => {};
