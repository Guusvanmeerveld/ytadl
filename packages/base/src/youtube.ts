import axios from 'axios';
import ytdl, { Filter } from 'ytdl-core';
import { parseStringPromise } from 'xml2js';

import { streamOptions } from './config';

import { Entry, YoutubeFeed } from './interfaces/youtube';
import { FeedItem } from './interfaces/feed';

export const findChannelIdByName = async (name: string): Promise<string | void> => {
	const { data, status } = await axios.get(`https://youtube.com/c/${name}`, {
		validateStatus: (status) => [200, 404].includes(status),
	});

	if (status == 404) {
		throw new Error(`Could not a find channel with the name '${name}'`);
	}

	const match =
		data.match(/"externalId":"([\w-]+)"/) || data.match(/channel-external-id="([\w-]+)"/);

	if (match) {
		return match[1];
	}
};

export const fetchFeedById = async (channelId: string): Promise<FeedItem[]> => {
	const { data, status } = await axios.get('https://www.youtube.com/feeds/videos.xml', {
		params: {
			channel_id: channelId,
		},
		validateStatus: (status) => [200, 404].includes(status),
	});

	if (status == 404) {
		throw new Error(`Could not a find channel with the id '${channelId}'`);
	}

	const json: YoutubeFeed = await parseStringPromise(data);

	const feed = json.feed.entry;

	return feed.map(YtItemtoFeedItem);
};

const YtItemtoFeedItem = (item: Entry): FeedItem => {
	const url = item.link.map((item) => item.$.href)[0];

	return {
		channel: item.author.map((item) => item.name.join(''))[0],
		platform: 'youtube',
		id: item['yt:videoId'][0],
		title: item.title[0],
		url,
		stream: async (options) => {
			const parsed = streamOptions.parse(options);

			let filter: Filter = 'audioandvideo';

			if (!(parsed?.audio?.included && parsed?.video?.included)) {
				if (parsed?.audio?.included) {
					filter = 'audioonly';
				}

				if (parsed?.video?.included) {
					filter = 'videoonly';
				}
			}

			return ytdl(url, { filter });
		},
	};
};
