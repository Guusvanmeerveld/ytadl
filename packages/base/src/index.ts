import { EventEmitter } from 'events';

import { z } from 'zod';

import schedule from 'node-schedule';

import ConfigSchema from './config';

import { fetchFeedById, findChannelIdByName } from './youtube';

import Emitter from './interfaces/emitter';
import { FeedItem } from './interfaces/feed';
import { Config, ConfigItem } from './interfaces/config';

export default class ytadl extends (EventEmitter as Emitter) {
	private data: Record<string, string[]> = {};
	private schema: z.output<typeof ConfigSchema>;

	/**
	 * Creates a new ytadl listener
	 * @param schema The input schema, an array of objects with each a name property with the channel name to listen for.
	 * @param config An optional config containing tokens or general settings
	 */
	constructor(schema: z.input<typeof ConfigSchema>, private config?: Config) {
		super();

		this.schema = ConfigSchema.parse(schema);

		this.init();
	}

	private init = async () => {
		this.schema = await Promise.all(
			this.schema.map(async (item) => {
				if (item.id || !item.name) return item;

				switch (item.platform) {
					case 'youtube':
						const id = await findChannelIdByName(item.name);

						if (!id) return item;

						return { ...item, id };
				}

				return item;
			})
		);

		for (const item of this.schema) {
			const job = schedule.scheduleJob(item.cron, async () => {
				if (!item.id) return;

				const [ids, feed] = await this.fetch(item);

				const difference = this.data[item.id].filter((i) => !ids.includes(i));

				this.data[item.id] = ids;

				if (difference.length > 0) {
					difference.forEach((feedItem) =>
						this.emit('newItem', feed.find(({ id }) => id == feedItem) as FeedItem, item)
					);
				}
			});

			if (!item.id) return;

			const [ids] = await this.fetch(item);

			this.data[item.id] = ids;

			job.on('run', () => this.emit('jobRan', item));
		}
	};

	private fetch = async (item: ConfigItem): Promise<[string[], FeedItem[]]> => {
		if (!item.id) throw new Error('Channel does not have an id');

		let feed: FeedItem[] | undefined;

		switch (item.platform) {
			case 'youtube':
				feed = await fetchFeedById(item.id);
				break;
		}

		if (!feed) throw new Error('Could not fetch remote feed');

		return [feed.map((i) => i.id), feed];
	};
}
