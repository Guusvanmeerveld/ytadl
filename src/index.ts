import { EventEmitter } from 'events';

import { z } from 'zod';

import schedule from 'node-schedule';

import ConfigSchema, { configItem } from './config';

import { fetchFeedById, findChannelIdByName } from './youtube';

import TypedEmitter from './interfaces/emitter';
import { FeedItem } from './interfaces/feed';

type ConfigItem = z.infer<typeof configItem>;

type EmitMap = {
	newItem: (item: FeedItem) => void;
	jobRan: (config: ConfigItem) => void;
};

const DEFAULT_CRON = '0 * * * *';

export default class ytadl extends (EventEmitter as new () => TypedEmitter<EmitMap>) {
	private data: Record<string, string[]> = {};

	constructor(private config: z.infer<typeof ConfigSchema>) {
		super();

		this.config = ConfigSchema.parse(config);

		this.init();
	}

	private init = async () => {
		this.config = await Promise.all(
			this.config.map(async (item) => {
				if (item.id || !item.name) return item;

				switch (item.platform) {
					case 'youtube':
						const id = await findChannelIdByName(item.name);

						if (!id) return item;

						return { ...item, id };
				}

				return item;
			})
		);

		for (const item of this.config) {
			const job = schedule.scheduleJob(item.cron ?? DEFAULT_CRON, async () => {
				if (!item.id) return;

				const [ids, feed] = await this.fetch(item);

				const difference = this.data[item.id].filter((i) => !ids.includes(i));

				if (difference.length > 0) {
					difference.forEach((feedItem) => {
						this.emit('newItem', feed.find(({ id }) => id == feedItem) as FeedItem);
					});
				}
			});

			if (!item.id) return;

			const [ids] = await this.fetch(item);

			this.data[item.id] = ids;

			job.on('run', () => this.emit('jobRan', item));
		}
	};

	private fetch = async (item: ConfigItem): Promise<[string[], FeedItem[]]> => {
		if (!item.id) throw new Error('Channel does not have an id');

		let feed: FeedItem[] | undefined;

		switch (item.platform) {
			case 'youtube':
				feed = await fetchFeedById(item.id);
				break;
		}

		if (!feed) throw new Error('Could not fetch remote feed');

		return [feed.map((i) => i.id), feed];
	};
}
