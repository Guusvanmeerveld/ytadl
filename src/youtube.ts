import axios from 'axios';
import ytdl, { Filter } from 'ytdl-core';
import { parseStringPromise } from 'xml2js';

import { streamOptions } from './config';

import { Entry, YoutubeFeed } from './interfaces/youtube';
import { FeedItem } from './interfaces/feed';

export const findChannelIdByName = async (name: string): Promise<string | void> => {
	const { data } = await axios.get<string>(`https://youtube.com/c/${name}`);

	const match =
		data.match(/"externalId":"([\w-]+)"/) || data.match(/channel-external-id="([\w-]+)"/);

	if (match) {
		return match[1];
	}
};

export const fetchFeedById = async (channelId: string): Promise<FeedItem[]> => {
	const { data } = await axios.get('https://www.youtube.com/feeds/videos.xml', {
		params: {
			channel_id: channelId,
		},
	});

	const json: YoutubeFeed = await parseStringPromise(data);

	const feed = json.feed.entry;

	return feed.map(YtItemtoFeedItem);
};

const YtItemtoFeedItem = (item: Entry): FeedItem => {
	const url = item.link.map((item) => item.$.href)[0];

	return {
		channel: item.author.map((item) => item.name.join(''))[0],
		platform: 'youtube',
		id: item['yt:videoId'][0],
		title: item.title[0],
		url,
		stream: async (options) => {
			const parsed = streamOptions.parse(options);

			let filter: Filter = 'audioandvideo';

			if (!(parsed?.audio?.included && parsed?.video?.included)) {
				if (parsed?.audio?.included) {
					filter = 'audioonly';
				}

				if (parsed?.video?.included) {
					filter = 'videoonly';
				}
			}

			return ytdl(url, { filter });
		},
	};
};
