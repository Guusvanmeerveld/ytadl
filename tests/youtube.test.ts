import { fetchFeedById, findChannelIdByName } from '../src/youtube';

const exampleChannelName = 'Fireship';
const exampleChannelId = 'UCsBjURrPoezykLs9EqgamOA';

describe('Test fetching Youtube videos', () => {
	test('It returns a Youtube channel id', async () => {
		const actual = await findChannelIdByName(exampleChannelName);

		expect(actual).toBe(exampleChannelId);
	});

	test("It throws an error because the channel name can't be found", async () => {
		expect.assertions(1);
		return findChannelIdByName('This channel does not exist').catch((e) =>
			expect(e).toMatchSnapshot()
		);
	});

	test("It throws an error because the channel id can't be found", async () => {
		expect.assertions(1);
		return fetchFeedById('This channel id does not exist').catch((e) =>
			expect(e).toMatchSnapshot()
		);
	});

	test('It returns a list of Youtube videos', async () => {
		const actual = await fetchFeedById(exampleChannelId);

		expect(actual).toBeDefined();
	});

	test('It returns a readable stream', async () => {
		const items = await fetchFeedById(exampleChannelId);

		const actual = await items[0].stream();

		expect(actual).toBeDefined();
	});
});
