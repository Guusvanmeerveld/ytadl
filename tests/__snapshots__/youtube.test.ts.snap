// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test fetching Youtube videos It throws an error because the channel id can't be found 1`] = `[Error: Could not a find channel with the id 'This channel id does not exist']`;

exports[`Test fetching Youtube videos It throws an error because the channel name can't be found 1`] = `[Error: Could not a find channel with the name 'This channel does not exist']`;
